File: src_matlab/chordal_extension.m
function [cliques, sizes, newA] = chordal_extension(A, turn)
%% Chordal extension of a graph
%%
%
% INPUT:
%   A: weight matrix or adjacency matrix (double)
%   turn: if 'on' draw the chordal graph (str)
%
% OUTPUT:
%   cliques: set of maximal cliques in the chordal graph (cell)
%   sizes: sizes of each clique in the chordal graph (double)
%
%% Author: T. Chen
%%
dense = 10;
n = size(A,1);
if isequal(turn, 'on')
    subplot(1,2,1)
    G = graph;
    for i = 1:n-1
        for j = i+1:n
            if A(i,j) ~= 0
                G = addedge(G, i, j);
            end
        end
    end
    if ismultigraph(G)
        G = simplify(G);
    end
    plot(G)
    title(sprintf('Original Graph: %d vertices, %d edges', numnodes(G), numedges(G)))
    axis('square')
end
A = spones(A) + (2*n+1)*speye(n);
opts.dense=dense;
p = amd(A, opts);
ch = chol(A(p, p));
clique = spones(ch);
[~, OrigIdx] = sort(p);
RemainIdx = 1;
for i = 2:n
    idx = i:n;
    num = find(clique(i, idx));
    NumOne = length(num);
    CliqueResult = sum(clique(RemainIdx,idx(num)), 2);
    if isempty(find(CliqueResult == NumOne, 1))
        RemainIdx = [RemainIdx; i];
    end
end
set = clique(RemainIdx, OrigIdx);
[elem,~] = find(set');
NoElem = full(sum(set, 2));
cliques = cell(length(NoElem), 1); sizes = [];
for i = 1:length(NoElem)
    cliques{i} = elem(sum(NoElem(1:i-1)) + 1:sum(NoElem(1:i)))';
    sizes = [sizes, NoElem(i)];
end
% draw the chordal graph
G = graph;
for i = 1:length(NoElem)
    for j = 1:NoElem(i)-1
        for k = j+1:NoElem(i)
            G = addedge(G, elem(sum(NoElem(1:i-1))+j), elem(sum(NoElem(1:i-1))+k));
        end
    end
end
if ismultigraph(G)
    G = simplify(G);
end
newA = adjacency(G);
if isequal(turn, 'on')
    subplot(1,2,2)
    plot(G)
    title(sprintf('Chordal Graph: %d vertices, %d edges', numnodes(G), numedges(G)))
    axis('square')
end
end

File: src_julia/chordal_extension.jl
using AMD
using LinearAlgebra
using MAT
using SparseArrays

function chordal_extension(A)
    lb = size(A, 1);
    A = sparse(1*(A.!=0)) + (2*lb + 1)*sparse(I, lb, lb);
    p = amd(A);
    R = cholesky(Matrix(A[p, p]));
    Cliques = sparse(1*(R.U.!=0));
    org_idx = sortperm(p);
    remainIdx = [1];
    for i = 2:11
        idx = i:lb;
        num = findall(Cliques[i, idx].!=0);
        noOfone = length(num);
        cliqueResult = sum(Cliques[remainIdx, idx[num]], dims = 2);
        if isempty(findall(cliqueResult.==noOfone))
            push!(remainIdx, i);
        end
    end
    cSet = Cliques[remainIdx, org_idx];
    Elem = [];
    for i = 1:size(cSet, 1)
        Elem = vcat(Elem, findall(cSet[i,:].!=0));
    end
    NoElem = Matrix(sum(cSet, dims = 2));
    cliques = Array{Any}(undef, length(NoElem));
    sizes = [];
    for i = 1:length(NoElem)
        cliques[i] = Elem[sum(NoElem[1:i-1])+1:sum(NoElem[1:i])]';
        push!(sizes, NoElem[i]);
    end
    return cliques, sizes
end

vars = matread("A.mat");
A = vars["A"];
c, s = chordal_extension(A);
